src/main/sensors/compass.c

```diff
@@ -402,36 +422,160 @@ uint32_t compassUpdate(timeUs_t currentTimeUs)
     }
 
     flightDynamicsTrims_t *magZero = &compassConfigMutable()->magZero;
-    if (magInit) {              // we apply offset only once mag calibration is done
-        mag.magADC[X] -= magZero->raw[X];
-        mag.magADC[Y] -= magZero->raw[Y];
-        mag.magADC[Z] -= magZero->raw[Z];
-    }
 
     if (tCal != 0) {
-        if ((currentTimeUs - tCal) < 30000000) {    // 30s: you have 30s to turn the multi in all directions
+        if (cmpTimeUs(tCal, currentTimeUs) > 0) {
             LED0_TOGGLE;
-            for (int axis = 0; axis < 3; axis++) {
-                if (mag.magADC[axis] < magZeroTempMin.raw[axis])
-                    magZeroTempMin.raw[axis] = mag.magADC[axis];
-                if (mag.magADC[axis] > magZeroTempMax.raw[axis])
-                    magZeroTempMax.raw[axis] = mag.magADC[axis];
+
+            // it is assumed that the user has started to move the quad if squared norm of rotational speed vector is greater than GYRO_NORM_SQUARED_MIN
+            float gyroNormSquared = 0.0f;
+            for (int axis = 0; axis < XYZ_AXIS_COUNT; axis++) {
+                gyroNormSquared += sq(DEGREES_TO_RADIANS(gyroGetFilteredDownsampled(axis)));
+            }
+
+            // check if movement has started
+            if (!didMovementStart && gyroNormSquared > GYRO_NORM_SQUARED_MIN) {
+                didMovementStart = true;
+                // starting now, the user has CALIBRATION_TIME_US to move the quad in a figure of eight in all directions
+                tCal = micros() + CALIBRATION_TIME_US;
+            }
+
+            // only start calibration after the user has started to move the quad
+            if (didMovementStart) {
+                compassBiasEstimatorApply(&compassBiasEstimator, mag.magADC);
             }
         } else {
             tCal = 0;
-            for (int axis = 0; axis < 3; axis++) {
-                magZero->raw[axis] = (magZeroTempMin.raw[axis] + magZeroTempMax.raw[axis]) / 2; // Calculate offsets
-            }
 
+            // copy the estimated bias and save it to the config
+            for (int axis = 0; axis < XYZ_AXIS_COUNT; axis++) {
```

Comment:
Maybe skip offset update (and save) if `didMovementStart` is false (calibration wait timeout).
In future, it may be also good idea to test if calibration succeeded

Reply:
Good point, thx